design-proposals/nic-hotplug/nic-hotplug.md

```diff
@@ -0,0 +1,359 @@
+Hot-plug / hot-unplug (add / remove) nics to running VMs is an industry
+standard available in multiple platforms, allowing the dynamic attachment of L2
+networks. This is useful when the workload (VM) cannot tolerate a restart when
+attaching / removing networks, of for scenarios where, for instance, the
+workload is created prior to the network.
+
+This would help bridge the gap between KubeVirt's feature set and what the user
+base expects from it.
+
+## Goals
+Adding network interfaces to running VMs.
+Removing networking interfaces from running VMs.
+A VM can have multiple interfaces connected to the same (secondary) network(s).
+
+## Non Goals
+- Providing multiple connections to the primary cluster network (managed by
+Kubernetes).
+- Dynamic attachment / removal of SR-IOV (or macvtap) networks; since these rely
+on a device plugin - and thus require the update of the pod's resources - i.e.
+request the VF resource - something which, is not possible on a running pod.
+This particular goal requires more study, and should be discussed in a separate
+design document.
+
+## Definition of Users
+This feature is intended for cluster users who want to connect their existing
+Virtual Machines to additional existing networks.
+These users may - or may not - have permissions to create the aforementioned
+additional networks; if so, they would need to rely on the cluster
+administrator to provision it for them.
+
+## User Stories
+* as a user, I want to add a new network interface to a running VM.
+* as a user, I want to remove a new network interface from a running VM.
+
+## Repos
+- [CNAO](https://github.com/kubevirt/cluster-network-addons-operator)
+- [KubeVirt](https://github.com/kubevirt/kubevirt)
+- [Multus](https://github.com/k8snetworkplumbingwg/multus-cni)
+
+# Design
+
+## Multus
+Multus - a CNI plugin - only handles the ADD / REMOVE verb, and is triggered
+by kubelet only when the pod's sandbox is created - or removed. Given its
+simplicity, it assumes no networks exist whenever it is executed, and procceeds
+to call the ADD / DEL for **all** networks listed in its
+`k8s.v1.cni.cncf.io/networks` annotation.
+
+As such, the first part of the design implementation must focus on multus; it
+must be refactored to work as a thick plugin, which enables it to be triggered
+not only when the pod's sandbox is created, but also on-demand.
+
+### Thickening multus
+A "thin CNI plugin" runs as a one-shot process, typically as a binary on disk
+executed on a Kubernetes host machine.
+
+A "thick CNI Plugin", on the other hand, is a CNI component composed of two (or
+more) parts, usually composed of "shim", and a long lived process (daemon)
+resident in memory. The "shim" is a lightweight "thin CNI plugin" component that
+simply passes CNI parameters (such as JSON configuration, and environment
+variables) to the daemon component, which then processes the CNI request.
+
+To transform multus into a thick plugin, it is needed to instantiate a long
```

Comment:
Unsure where else to comment this, but... something I'd like to have done as part of the thickening is to use a more-standard authentication to k8s api via client-go rather than what's in the current thin plugin -- the generated KUBECONFIG. This sometimes causes intermittent auth failures which flake out CI. I think if we rely on the service account auth tokens and use client-go, we'll clear that up.